src/heightmap.cpp

```diff
@@ -22,6 +22,47 @@
 
 #include "safeguards.h"
 
+/**
+ * Maximum number of pixels for one dimension of a heightmap image.
+ *
+ * At some point the number of pixels in a heightmap becomes larger compared to
+ * the pixels that will be taken for the map. With this limit already 75% of the
+ * pixels would be discarded for the largest image and the largest map size.
+ */
+static const uint MAX_HEIGHTMAP_SIDE_LENGTH_IN_PIXELS = 2 * MAX_MAP_SIZE;
+
+/*
+ * Maximum size, in bytes, of the heightmap to consider loading.
+ *
+ * The theoretical maximum size of an image is about 16 exabytes for PNG and
+ * about 4 exabytes for a BMP image. On a 32 bit platform this will overflow
+ * size_t which could lead to memory corruption.
+ * Even on a 64 bit platform that is way larger than feasible; we could allow it
+ * on those platforms, but then the memory allocation that is done will almost
+ * certainly fail causing a crash to the desktop instead of a nice user error.
+ * So, artificially limit the maximum size of a hightmap image to allocate one
+ * gigabyte here.
+ */
+static const uint MAX_HEIGHTMAP_SIZE_IN_BYTES = 1 << 30;
```

Comment:
`Ensure memory is within reason while loading a heightmap.`

Not sure if the rest really added value for me. It kinda only confused, as it sounded like it is much more than it really is, if I understand you correctly now.

Reply:
More like: "why are you attempting to load a more than 256 million pixel image as a heightmap? You would be discarding a lot of pixels in the original image!"

The maximum number of tiles for vanilla us is 16 million. Since the mapping from image heightmap to tile map iterates over the tiles and then gets the height data from the associated image pixel, any pixel that is not mapped from a tile will not even be used. So, that would mean that more than 95% of the pixels will only be loaded but never used in creating the heights of the tile map.
Practically the new current side length limit of 8kx8k would be 64 million pixels would not need the 256M limit. However, the code that existed was explicitly there to prevent overflows when loading really large heightmaps, so if in the future we were to increase the maximum number of tiles along an edge, that overflowing issue might return. As such the limit on the total number of pixels. That limit would already be used by JGRPP, so it's not actually pointless right now.